### PR TITLE
chore: handle nans in GPU training

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: python/python/lance/torch/distance.py [hunk 1]
```diff
@@ -232,4 +234,5 @@ def dot_distance(x: torch.Tensor, y: torch.Tensor) -> Tuple[torch.Tensor, torch.
     dists = 1 - x @ y.T
     idx = torch.argmin(dists, dim=1, keepdim=True)
     dists = dists.take_along_dim(idx, dim=1).reshape(-1)
+    idx = torch.where(dists.isnan(), torch.nan, idx)
```
Comment (westonpace) [inline python/python/lance/torch/distance.py:237]: What is this line doing?  It looks like you're replacing nan with nan.  Is `torch.nan` somehow different than normal nan?

File: python/python/lance/vector.py [hunk 2]
```diff
@@ -215,26 +215,37 @@ def compute_partitions(
...
+                    partitions = partitions[partitions != -1]
+                    part_batch = pa.RecordBatch.from_arrays(
+                        [ids, partitions],
                         schema=output_schema,
                     )
+                    if len(part_batch) < len(ids):
+                        logging.warning(
+                            "%s vectors are ignored during partition assignment",
+                            len(part_batch) - len(ids),
+                        )
```
Comment (westonpace) [inline python/python/lance/vector.py:242]: What happens to these vectors?  Do they keep their old partition id?

**Review by westonpace — APPROVED** [review]
Minor questions.  Also, I appreciate the improvements to method signatures and names!

File: python/python/lance/torch/distance.py (hunk 1 above)
Comment (eddyxu) [reply python/python/lance/torch/distance.py:237]: This line sets lines in idx array based on the `nan` value from dist array.

File: python/python/lance/vector.py (hunk 2 above)
Comment (eddyxu) [reply python/python/lance/vector.py:242]: These vectors are not stored in the vector index (still in the dataset tho).

File: python/python/lance/vector.py [hunk 3]
```diff
@@ -215,26 +217,37 @@ def compute_partitions(
...
         )
 
-        def _f() -> Iterable[pa.RecordBatch]:
+        def _partition_assignment() -> Iterable[pa.RecordBatch]:
             with torch.no_grad():
                 for batch in torch_ds:
                     batch: Dict[str, torch.Tensor] = batch
                     vecs = batch[column].reshape(-1, kmeans.centroids.shape[1])
-                    ids = batch["_rowid"].reshape(-1)
+                    ids = batch["_rowid"].reshape(-1).cpu().numpy()
```
Comment (chebbyChefNEQ) [inline python/python/lance/vector.py:225]: can we move this to after we call `kmeans.transform` to get some pipeing?

File: rust/lance-index/src/vector/ivf.rs [hunk 4]
```diff
@@ -463,23 +462,11 @@ impl<T: ArrowFloatType + Dot + L2 + Cosine + 'static> Ivf for IvfImpl<T> {
...
                         builder.append_value(*part_id);
                     } else {
-                        return Err(Error::Index {
-                            message: format!(
-                                "Row ID {} does not exist in the precomputed partitions",
-                                row
-                            ),
-                            location: location!(),
-                        });
+                        builder.append_null();
```
Comment (chebbyChefNEQ) [inline rust/lance-index/src/vector/ivf.rs:469]: do we need to handle the null in the shuffle?

Comment (eddyxu) [reply rust/lance-index/src/vector/ivf.rs:469]: These null partitions will be filtered out.

File: python/python/lance/vector.py (hunk 3 above)
Comment (eddyxu) [reply python/python/lance/vector.py:225]: Done

**Review by chebbyChefNEQ — APPROVED** [review]
